openrndr-svg/src/main/kotlin/org/openrndr/svg/SVGElement.kt

```diff
@@ -0,0 +1,394 @@
+
+            if (result.size >= 2) {
+                styleProperty(result[0], result[1])
+            }
+        }
+    }
+}
+
+/** <svg> element */
+internal class SVGSVGElement(element: Element): SVGGroup(element), SVGFitToViewBox, SVGDimensions {
+    override var viewBox: Rectangle? = SVGParse.viewBox(this.element)
+    override var preserveAspectRatio: Alignment = SVGParse.preserveAspectRatio(this.element)
+
+    override var bounds = SVGParse.bounds(this.element)
+    /** Represents the scale and translate applied to the viewport */
+    override var currentTransform = calculateViewportTransform()
+
+    /**
+     * Calculates effective viewport transformation using [viewBox] and [preserveAspectRatio].
+     * As per [the SVG 2.0 spec](https://svgwg.org/svg2-draft/single-page.html#coords-ComputingAViewportsTransform)
+     */
+    fun calculateViewportTransform(): Matrix44 {
+        return when {
+            viewBox != null -> {
+                // TODO! Someone tell me how to shorten this
+                val vbCorner = viewBox!!.corner
+                val vbDims = viewBox!!.dimensions
+                // TODO! Do we need to know DPI at this point?
+                // Should this function be in Composition.kt instead?
+                val eCorner = bounds.position.vector2
+                val eDims = bounds.dimensions.vector2
+                val (xAlign, yAlign, meetOrSlice) = preserveAspectRatio
+
+                var scale = eDims / vbDims
+
+                if (xAlign != Align.NONE && yAlign != Align.NONE) {
+                    scale = if (meetOrSlice == MeetOrSlice.MEET) {
+                        Vector2(min(scale.x, scale.y))
+                    } else {
+                        Vector2(max(scale.x, scale.y))
+                    }
+                }
+
+                var translate = eCorner - (vbCorner * scale)
+
+                translate = when (xAlign) {
+                    Align.MID -> translate.copy(x = translate.x + (eDims.x - vbDims.x * scale.x) / 2)
+                    Align.MAX -> translate.copy(x = translate.x + (eDims.x - vbDims.x * scale.x))
+                    else -> translate
+                }
+
+                translate = when (yAlign) {
+                    Align.MID -> translate.copy(y = translate.y + (eDims.y - vbDims.y * scale.y) / 2)
+                    Align.MAX -> translate.copy(y = translate.y + (eDims.y - vbDims.y * scale.y))
+                    else -> translate
+                }
+
+                buildTransform {
+                    translate(translate)
+                    scale(scale.x, scale.y, 1.0)
+                }
+            }
+            viewBox == Rectangle.EMPTY -> {
+                // TODO! Questionable return lmao
+                // The intent is to not display the element
+                Matrix44.ZERO
+            }
+            else -> {
+                Matrix44.IDENTITY
+            }
+        }
+    }
+}
```

Comment:
Is it possible to make this more readable? Those translate calculations look cumbersome